firmware/controller-128-no-clock/controller-128-no-clock.ino

```diff
@@ -0,0 +1,835 @@
+
+  void tick() {
+    ticks++;
+    if (ticks >= duration) {
+      step();
+    }
+  }
+
+  void step() {
+    ticks = 0;
+    if (arp == ARP_FORWARD) {
+      current++;
+      if (current >= length) {
+        current = 0;
+      }
+    }
+    else if (arp == ARP_REVERSE) {
+      current--;
+      if (current <= start) {
+        current = length - 1;
+      }
+    }
+    // todo: add'l arp ...
+  }
+
+  void reset() {
+    ticks = 0;
+    current = phase;
+  }
+
+  boolean get() {
+    return bits[start + current];
+  }
+
+  boolean get(int i) {
+    return bits[start + i];
+  }
+
+  void set(int i, boolean v) {
+    bits[start + i] = v;
+  }
+
+  void toggle(int i) {
+    if (get(i)) {
+      set(i, false);
+    }
+    else {
+      set(i, true);
+    }
+  }
+
+  void clear() {
+    for (int i = 0; i < length; i++) {
+      set(i, false);
+    }
+  }
+
+  boolean check(int i) {
+    return (i >= start) && (i < (start + length));
+  }
+
+  boolean highlighted(int i) {
+    return i == (current + start);
+  }
+
+  boolean selected(int i) {
+    return check(i) && get(i);
+  }
+
+  boolean pressed(int i) {
+    return false;    
+  }
+
+  boolean deselected(int i) {
+    return check(i);
+  }
+};
+
+Track tracks[7];
+
+
+//
+// buttons
+//
+
+// top row buttons
+#define CLOCK_BUTTON 0
+#define RESET_BUTTON 1
+#define CLEAR_BUTTON 2
+#define INCREASE_START_BUTTON 3 // or cycle start?
+#define DECREASE_START_BUTTON 4
+#define INCREASE_PHASE_BUTTON 5
+#define DECREASE_PHASE_BUTTON 6
+#define INCREASE_LENGTH_BUTTON 7
+#define DECREASE_LENGTH_BUTTON 8
+#define INCREASE_DURATION_BUTTON 9
+#define DECREASE_DURATION_BUTTON 10
+#define INCREASE_ARP_BUTTON 11
+#define DECREASE_ARP_BUTTON 12
+#define FOCUS_BUTTON 13
+#define INCREASE_FOCUS_BUTTON 14 // or page right?
+#define DECREASE_FOCUS_BUTTON 15 // or page left
+
+// focus
+int focus_index = 0;
+boolean focus_mode = false;
+
+// cache trigger values
+boolean trigger1 = false;
+boolean trigger2 = false;
+boolean trigger3 = false;
+boolean trigger4 = false;
+boolean trigger5 = false;
+boolean trigger6 = false;
+boolean trigger7 = false;
+
+void clock_pressed() {
+  trellis.setPixelColor(CLOCK_BUTTON, 0, pressed);
+  trellis.show();
+}
+
+void clock_released() {
+  trellis.setPixelColor(CLOCK_BUTTON, 0, off);
+  trellis.show();
+
+  // tick
+  for (int i = 0; i < 7; i++) {
+    Track* t = &tracks[i];
+    t->tick();
+    if (t->get()) {
+      switch (i) {
+        case 0: digitalWrite(TRIGGER_1, HIGH); trigger1 = true; break;
+        case 1: digitalWrite(TRIGGER_2, HIGH); trigger2 = true; break;
+        case 2: digitalWrite(TRIGGER_3, HIGH); trigger3 = true; break;
+        case 3: digitalWrite(TRIGGER_4, HIGH); trigger4 = true; break;
+        case 4: digitalWrite(TRIGGER_5, HIGH); trigger5 = true; break;
+        case 5: digitalWrite(TRIGGER_6, HIGH); trigger6 = true; break;
+        case 6: digitalWrite(TRIGGER_7, HIGH); trigger7 = true; break;
+        default: break;
+      }
+    }
+  }
```

Comment:
Wouldn't you want to do this in clock_pressed(), then set all triggers low on clock_released(), so that if you have two of the same trigger in a row, it triggers twice?

Reply:
Maybe so 🤔